tests/transforms/test_batch_transform.py

```diff
@@ -0,0 +1,427 @@
+        assert tapes[0].operations[1].name == "RY"
+        assert tapes[0].operations[1].parameters == [a * np.abs(x)]
+
+        assert len(tapes[1].operations) == 2
+        assert tapes[1].operations[0].name == "Hadamard"
+        assert tapes[1].operations[1].name == "RZ"
+        assert tapes[1].operations[1].parameters == [b * np.sin(x)]
+
+        expected = fn(dev.batch_execute(tapes))
+        assert res == expected
+
+
+@pytest.mark.parametrize("diff_method", ["parameter-shift", "backprop", "finite-diff"])
+class TestQFuncTransformGradients:
+    """Tests for the batch_transform decorator differentiability"""
+
+    @staticmethod
+    @qml.batch_transform
+    def my_transform(tape, weights):
+        """Generates two tapes, one with all RX replaced with RY,
+        and the other with all RX replaced with RZ."""
+
+        tape1 = qml.tape.JacobianTape()
+        tape2 = qml.tape.JacobianTape()
+
+        # loop through all operations on the input tape
+        for op in tape.operations + tape.measurements:
+            if op.name == "RX":
+                wires = op.wires
+                param = op.parameters[0]
+
+                with tape1:
+                    qml.RY(weights[0] * qml.math.sin(param), wires=wires)
+
+                with tape2:
+                    qml.RZ(weights[1] * qml.math.cos(param), wires=wires)
+            else:
+                for t in [tape1, tape2]:
+                    with t:
+                        qml.apply(op)
+
+        def processing_fn(results):
+            return qml.math.sum(qml.math.stack(results))
+
+        return [tape1, tape2], processing_fn
+
+    @staticmethod
+    def circuit(x):
+        """Test ansatz"""
+        qml.Hadamard(wires=0)
+        qml.RX(x, wires=0)
+        return qml.expval(qml.PauliX(0))
+
+    @staticmethod
+    def expval(x, weights):
+        """Analytic expectation value of the above circuit qfunc"""
+        return np.cos(weights[1] * np.cos(x)) + np.cos(weights[0] * np.sin(x))
+
+    def test_differentiable_autograd(self, diff_method):
+        """Test that a batch transform is differentiable when using
+        autograd"""
+        dev = qml.device("default.qubit", wires=2)
+        qnode = qml.QNode(self.circuit, dev, interface="autograd", diff_method=diff_method)
+
+        def cost(x, weights):
+            return self.my_transform(qnode, weights)(x)
+
+        weights = np.array([0.1, 0.2], requires_grad=True)
+        x = np.array(0.543, requires_grad=True)
+
+        res = cost(x, weights)
+        assert np.allclose(res, self.expval(x, weights))
+
+        grad = qml.grad(cost)(x, weights)
+        expected = qml.grad(self.expval)(x, weights)
+        assert all(np.allclose(g, e) for g, e in zip(grad, expected))
+
+    def test_differentiable_tf(self, diff_method):
+        """Test that a batch transform is differentiable when using
+        TensorFlow"""
+        if diff_method in ("parameter-shift", "finite-diff"):

```

Comment:
I thought tf supported parameter-shift? (I have this in some of my tests, e.g., `test_compile.py`.)

Reply:
Oh, this is just because the TF parameter-shift+batch support hasn't been merged into master yet 😄 #1542 

Once it is merged I can remove this if statement

Reply:
Ohhhh the difference is _batch_ support, I see.